src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,199 @@
+
+#define IST8310_MAG_I2C_ADDRESS 0x0C
+
+/* ist8310 Slave Address Select : default address 0x0C
+ *        CAD1  |  CAD0   |  Address
+ *    ------------------------------
+ *         VSS   |   VSS  |  0CH
+ *         VSS   |   VDD  |  0DH
+ *         VDD   |   VSS  |  0EH
+ *         VDD   |   VDD  |  0FH
+ * if CAD1 and CAD0 are floating, I2C address will be 0EH
+ *
+ *
+ * CTRL_REGA: Control Register 1
+ * Read Write
+ * Default value: 0x0A
+ * 7:4  0   Reserved.
+ * 3:0  DO2-DO0: Operating mode setting
+ *        DO3  |  DO2 |  DO1 |  DO0 |   mode
+ *    ------------------------------------------------------
+ *         0   |   0  |  0   |  0   |   Stand-By mode
+ *         0   |   0  |  0   |  1   |   Single measurement mode
+ *                                       Others: Reserved
+ *
+ * CTRL_REGB: Control Register 2
+ * Read Write
+ * Default value: 0x0B
+ * 7:4  0   Reserved.
+ * 3    DREN : Data ready enable control:
+ *      0: disable
+ *      1: enable
+ * 2    DRP: DRDY pin polarity control
+ *      0: active low
+ *      1: active high
+ * 1    0   Reserved.
+ * 0    SRST: Soft reset, perform Power On Reset (POR) routine
+ *      0: no action
+ *      1: start immediately POR routine
+ *      This bit will be set to zero after POR routine
+ */
+
+#define IST8310_REG_DATA 0x03
+#define IST8310_REG_WAI 0x00
+#define IST8310_REG_WAI_VALID 0x10
+
+#define IST8310_REG_STAT1 0x02
+#define IST8310_REG_STAT2 0x09
+
+#define IST8310_DRDY_MASK 0x01
+
+// I2C Control Register
+#define IST8310_REG_CNTRL1 0x0A
+#define IST8310_REG_CNTRL2 0x0B
+#define IST8310_REG_AVERAGE 0x41
+#define IST8310_REG_PDCNTL 0x42
+
+// Parameter
+// ODR = Output Data Rate, we use single measure mode for getting more data.
+#define IST8310_ODR_SINGLE 0x01
+#define IST8310_ODR_10_HZ 0x03
+#define IST8310_ODR_20_HZ 0x05
+#define IST8310_ODR_50_HZ 0x07
+#define IST8310_ODR_100_HZ 0x06
+
+#define IST8310_AVG_16  0x24
+#define IST8310_PULSE_DURATION_NORMAL 0xC0
+
+#define IST8310_CNTRL2_RESET 0x01
+#define IST8310_CNTRL2_DRPOL 0x04
+#define IST8310_CNTRL2_DRENA 0x08
+
+static bool ist8310Init(magDev_t *magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    busDeviceRegister(dev);
+
+    // Init setting
+    bool ack = busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(6);
+    ack = ack && busWriteRegister(dev, IST8310_REG_PDCNTL, IST8310_PULSE_DURATION_NORMAL);
+    delay(6); 
+
+    busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_SINGLE);
+
+    return ack;
+}
+
+static bool ist8310Read(magDev_t * magDev, int16_t *magData)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    static uint8_t buf[6];
+    const int LSB2FSV = 3; // 3mG - 14 bit
+
+    static enum {
+        STATE_REQUEST_DATA,
+        STATE_FETCH_DATA,
+    } state = STATE_REQUEST_DATA;
+
+    switch (state) {
+        default:
+        case STATE_REQUEST_DATA:
+            busReadRegisterBufferStart(dev, IST8310_REG_DATA, buf, sizeof(buf));
```

Comment:
Problem is that this will fail if some other transaction is pending.
Maybe just :
```suggestion
            if (!busReadRegisterBufferStart(dev, IST8310_REG_DATA, buf, sizeof(buf))) return false;
```